### ISSUE TITLE
Feature proposal: plugin script injection

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#361]
We tried out GUI posts and user pages, but I wasn't really happy with them, and I don't think yall were either. Iframes are a difficult tool. They don't perform well and the put you in a box.

I really like the idea of user-published plugins though. SSB is so flexible with data schemas that it's really asking for user code to interpret the messages. Plus, being a hostless network encourages the idea of governance by the community, instead of by a benevolent upstream.

With that in mind I was wondering, what if we let users publish plugins that we evaled on the page? 

This would be a bit of a social experiment. It would be security-by-community.
People would audit each other's packages and sign off on their correctness, safety, and so on.
With content-hash links, the code can't change after its been published, so we don't have to worry about hosts altering the script. If some plugin causes problems, there's a very clear line of responsibility, and people can be held accountable.
Of course, viruses will get into the network, but minimizing that will just be part of the experiment!

To do this, we'd update Phoenix with tools to publish plugins, do code review, publish audit results, and to manage your active plugins. We might also have config profiles, so that you can switch between bundles of plugins.

For the code, we could put together a toolset to make sure plugins interact well with each other. This will include things like proper cleanup when turned off, and using GUIs for users to review changes before they're made. If a plugin isn't a good citizen of the toolset, then it shouldn't get signatures. But the idea is that a plugin can really do anything -- the Phoenix shell isn't going to prohibit it.

Too crazy? What do yall think?

**Comment by dominictarr** [comment]
yes I think this will be a very good way to explore an auditing workflow.

**Comment by dominictarr** [comment]
just because an idea may seem crazy right now, doesn't mean it will still seem like that when we add other pieces around it.

**Comment by pfraze** [comment]
We decided (in IRC) to avoid a runtime modules system for now, so that the package manager can have time to develop independently. This means that plugins are going to need to be prepackaged (eg with browserify) so devs are going to need to use external toolsets, and plugins in development will need to load from the filesystem. (My original plan was to avoid that with an in-phoenix editor.)

We don't need a perfect process right now, just one that works. I suggest we create a `~/.ssb/plugins` directory and have devs symlink js files into it. Phoenix will load and eval all the js files in that directory at pageload. 

The dev process will then be to create the output bundle, then `ln -s ./bundle.js ~/.ssb/plugins/myplug.js`. You can continue creating new output bundles, and refreshing phoenix to see changes. To stop loading the plugin, `rm ~/.ssb/plugins/myplug.js`.

Sound good?

**Comment by dominictarr** [comment]
:+1:
